lib/internal/warn.js

```diff
@@ -0,0 +1,60 @@
+'use strict';
+
+// A process Warning is similar in structure to an Error except that they
+// are emitted by the process.on('warn') event rather than being thrown.
+
+const _message = Symbol('message');
+const _name = Symbol('name');
+const prefix = `(${process.release.name}:${process.pid}) `;
+
+class Warning {
+  constructor(name, message, constructor) {
+    this[_name] = name;
+    this[_message] = `${prefix}${message}`;
+    Error.captureStackTrace(this, constructor || Warning);
+  }
+
+  get message() {
+    return this[_message];
+  }
+
+  get name() {
+    return this[_name];
+  }
+}
+
+function emit(warning) {
+  process.nextTick(() => {
+    process.emit('warn', warning);
+  });
+}
+
+Warning.deprecationWarning = function(message, warned) {
+  if (process.noDeprecation || warned) return true;
+  const warning = new Warning('DeprecationWarning',
```

Comment:
I kind of feel we're repeating DeprecationWarning here twice (first and third parameters)